### ISSUE TITLE
RLMCollectionChange notifies about modification with primary key change.

### ISSUE THREAD

**Issue by molind** [realm/realm-swift#4732]
## Goals

Receive changes notification from realm

## Expected Results

Primary key shouldn't change during modification.

## Actual Results

We queried sorted results and subscribed for notifications for those objects. During network sync. One of the notifications contains following changes:

* insert 0-17, 19-30
* modify 0
* delete - empty

It should be applied to array with 1 object with primary key "123"
We insert new data. And modify contents of object at index 18 (That's our initial object moved after insert). Assert triggered when we check difference between RLMResults and our array of objects. RLMResults already contains object with primary key "123" at index 14. And during modification we should replace contents of our object with contents of another object including primary key. That's confuses us a lot.

## Steps to Reproduce

It's happened when we connected to synced realm and received some part of transactions from it. Before sync in local realm we had same data, that stored on realm-object-server. Before sync we copied our data into remote realm using:

## Code Sample

```
// part of function that enables sync
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
            RLMRealm *curRealm = [RLMRealm defaultRealm];
            
            RLMRealmConfiguration *config = [[RLMRealmConfiguration alloc] init];
            config.syncConfiguration = [[RLMSyncConfiguration alloc] initWithUser:user realmURL:self.realmURL];
            NSError *error = nil;
            RLMRealm *newRealm = [RLMRealm realmWithConfiguration:config error:&error];
            if(newRealm != nil && error == nil)
            {
                try {
                    [newRealm beginWriteTransaction];
                    [self copyAllObjectFromRealm:curRealm toRealm:newRealm];
                    if ([newRealm commitWriteTransaction:&error]) {
                        dispatch_async(dispatch_get_main_queue(), ^{
                            _syncAuthRunning = NO;
                            [RLMRealmConfiguration setDefaultConfiguration:config];
                            [self updateRealmCallbacks];
                            [AppSettings settings].syncAuth = syncAuth;
                        });
                    }
                } catch (...) {
                    [self during:@"Commit transaction" happened:[[NSError alloc] initWithDomain:GalileoErrorDomain code:Error_RealmCommit userInfo:@{NSLocalizedDescriptionKey: @"Exception happened during commit transaction. Cloud sync is disabled."}]];
                }
            } else {
                [user logOut];
            }
            if (error) {
                [self during:@"Create sync realm" happened:error];
            }
        });
}}

// copies our data into remote realm during inital sync.
-(void) copyAllObjectFromRealm:(RLMRealm *)fromRealm toRealm:(RLMRealm *)toRealm
{
    for(ModelTrack *item in [ModelTrack allObjectsInRealm:fromRealm])
    {
        [ModelTrack createOrUpdateInRealm:toRealm withValue:item];
    }
}

// part of function that receives object changes.
RLMResults *tracks = [[ModelTrack objectsInRealm:[RLMRealm defaultRealm] withPredicate:predicate] sortedResultsUsingKeyPath:@"uuid" ascending:NO];
    _tracksNotificationToken = [tracks addNotificationBlock:^(RLMResults *results, RLMCollectionChange *change, NSError *error){
// apply changes
}];
```

## Version of Realm and Tooling

Realm version: 2.4.4

Xcode version: .8.2.1

iOS/OSX version: 10.2.1

Dependency manager + version: Cocoapods 1.2.0

**Comment by istx25** [comment]
Hi @molind. Thanks for writing in. Would you be able to provide us with a sample project that includes your model classes themselves as well as enough code to reproduce the problem when we build/run the project?

**Comment by molind** [comment]
Hi @istx25. We'll try to gather all our db related code into one project. Just to make it clear. Do you think it's bug, or it works as expected when modifies primary key of object?

**Comment by molind** [comment]
Ok, I created basic test project to demonstrate this issue. https://github.com/molind/RealmTests But, seems I stuck into another our issue we discovered before: https://github.com/realm/realm-mobile-platform/issues/112

**Comment by jpsim** [comment]
Thanks for taking the time to prepare this @molind. I'm going to put this on hold until we can run your repro case without being blocked by realm/realm-mobile-platform#112.
